llvm_passes/HipPrintf.cpp

```diff
@@ -18,158 +53,363 @@
 #include "llvm/Passes/PassBuilder.h"
 #include "llvm/Passes/PassPlugin.h"
 
+#include <vector>
+#include <string>
+#include <iostream>
+
+#define SPIRV_OPENCL_CONSTANT_AS 2
+#define SPIRV_OPENCL_GENERIC_AS 4
+#define SPIRV_OPENCL_PRINTF_FMT_ARG_AS SPIRV_OPENCL_CONSTANT_AS
+
+#define ORIG_PRINTF_FUNC_NAME "_hip_printf"
+#define ORIG_PRINT_STRING_FUNC_NAME "_cl_print_str"
+
+// The SPIR-V specification doesn't especially forbid %s
+// arguments that are not "literals" (in constant address
+// space as specified by OpenCL). Assume a SPIR-V printf
+// can print strings that are not moved to constant space
+// and no special functions is needed for their printout.
+// The SPIR-V implementations seen so far do not care about this.
+#define ASSUME_PRINTF_SUPPORTS_GLOBAL_STRING_ARGS 1
+
 using namespace llvm;
 
-class HipPrintfToOpenCLPrintfLegacyPass : public ModulePass {
-public:
-  static char ID;
-  HipPrintfToOpenCLPrintfLegacyPass() : ModulePass(ID) {}
+unsigned NumFormatSpecs(StringRef FmtString) {
+  return FmtString.count("%") - 2 * FmtString.count("%%");
+}
 
-  bool runOnModule(Module &M) override {
-    return false;
+// Strip address space casts, GEPs etc. towards the global string
+// literal and return it.
+GlobalVariable *findGlobalStr(Value *Arg) {
+
+  if (auto GEP = dyn_cast<GetElementPtrInst>(Arg)) {
+    Arg = GEP->getPointerOperand();
+  }
+  if (auto ASCast = dyn_cast<AddrSpaceCastInst>(Arg)) {
+    Arg = ASCast->getPointerOperand();
+  } else if (auto CE = dyn_cast<ConstantExpr>(Arg)) {
+    if (CE->getOpcode() == llvm::Instruction::AddrSpaceCast) {
+      Arg = CE->getOperand(0);
+    } else {
+      assert("Cannot handle the printf format string!" && false);
+    }
   }
```

Comment:
Put else branch and call llvm_unreachable here? In case of an unexpected instruction is encountered.